From ea6d9bd5e8d4883bb5d7765c6e4c83395cf3a9da Mon Sep 17 00:00:00 2001
From: Kevin Shaum <kevin.shaum@cpanel.net>
Date: Mon, 22 Jul 2013 13:26:04 -0500
Subject: [PATCH 3/3] 2.2_cpanel_fileprotect_suexec_httpusergroupallow

cpanel-specific patch for file protection allowing execution
by suexec as long as the uid is in the supplemental groups
---
 support/suexec.c | 33 +++++++++++++++++++++++++++++++--
 1 file changed, 31 insertions(+), 2 deletions(-)

diff --git a/support/suexec.c b/support/suexec.c
index d93b22f..29d7210 100644
--- a/support/suexec.c
+++ b/support/suexec.c
@@ -268,6 +268,25 @@ static void clean_env(void)
     environ = cleanenv;
 }
 
+static int is_group_member( const char *uname, gid_t gid ) {
+    struct group *g;
+    char **members;
+
+    g = getgrgid( gid );
+    if ( g == NULL ) {
+        log_err("crit: cannot get membership of group %d\n", gid);
+        exit(109);
+    }
+
+    members = g->gr_mem;
+    while ( *members != NULL ) {
+        if ( strcmp( *members, uname ) == 0 ) {
+            return 1;
+        }
+    }
+    return 0;
+}
+
 int main(int argc, char *argv[])
 {
     int userdir = 0;        /* ~userdir flag             */
@@ -279,6 +298,7 @@ int main(int argc, char *argv[])
     char *target_uname;     /* target user name          */
     char *target_gname;     /* target group name         */
     char *target_homedir;   /* target home directory     */
+    gid_t default_gid;      /* Default GID for target UID*/
     char *actual_uname;     /* actual user name          */
     char *actual_gname;     /* actual group name         */
     char *cmd;              /* command to be executed    */
@@ -483,6 +503,7 @@ int main(int argc, char *argv[])
     uid = pw->pw_uid;
     actual_uname = strdup(pw->pw_name);
     target_homedir = strdup(pw->pw_dir);
+    default_gid = pw->pw_gid;
     if (actual_uname == NULL || target_homedir == NULL) {
         log_err("failed to alloc memory\n");
         exit(126);
@@ -516,6 +537,14 @@ int main(int argc, char *argv[])
     }
 
     /*
+     * Error out if user UID is not a member of group GID.
+     */
+    if ((gid != default_gid) && !is_group_member(actual_uname, gid)) {
+	log_err("crit: user does not belong to group (%s/%d/%s)\n", actual_uname, gid, cmd);
+	exit(108);
+    }
+
+    /*
      * Change UID/GID here so that the following tests work over NFS.
      *
      * Initialize the group access list for the target user,
@@ -649,9 +678,9 @@ int main(int argc, char *argv[])
      * the name/group of the cwd or the program.
      */
     if ((uid != dir_info.st_uid) ||
-        (gid != dir_info.st_gid) ||
+        ((gid != dir_info.st_gid) && (dir_info.st_gid != httpgid)) ||
         (uid != prg_info.st_uid) ||
-        (gid != prg_info.st_gid)) {
+        ((gid != prg_info.st_gid) && (prg_info.st_gid != httpgid))) {
 #ifdef TRUSTED_USERS_SCRIPTS
         /*
          * Check if we're using a "shared" script
-- 
1.7.11.3

